### PR TITLE
use toJSON() property when it exists

### PR DESCRIPTION
**Pull request by simon-id**
### What does this PR do?

When converting objects and arrays, if the `toJSON()` method is present, call it and use it as result

### CODE REVIEW

File: src/convert.cpp [hunk 1]
```diff
@@ -56,6 +56,12 @@ ddwaf_object* to_ddwaf_object_object(
   int depth,
   bool lim
 ) {
+  Napi::Value toJSON = obj.Get("toJSON");
+  if (toJSON.IsFunction()) {
+    Napi::Value json = toJSON.As<Napi::Function>().Call(obj, {});
+    obj = json.ToObject();
+  }
```
Comment (simon-id) [inline src/convert.cpp:63]: maybe this should only be used in certain addresses only ? or if we should only check the ownProperty and not inherited props ? I'm a bit afraid of some prototype pollution here

File: test/index.js
Comment (simon-id) [inline test/index.js:1]: TODO: add crazy json and assert that result of convert and of JSON.stringify are the same ?

File: src/convert.cpp
Comment (simon-id) [inline src/convert.cpp:1]: TODO: implement it for other conversion function too like scalars, not just objects

Comment (simon-id) [reply src/convert.cpp:1]: turns out it's only a problem for objects, so only needed for objects and arrays

Comment (uurien) [reply src/convert.cpp:1]: added the changes also for arrays

File: test/index.js
Comment (uurien) [reply test/index.js:1]: I added some tests (I dunno if they are crazy enough).
The result of `JSON.stringify` and our approach are not going to be always the same. We are converting the functions to `ddwaf_object_invalid` and `JSON.stringify` is just ignoring them

File: src/convert.cpp (hunk 1 above)
Comment (simon-id) [reply src/convert.cpp]: turns out JSON.stringify also uses the prototype so we're good

File: src/convert.cpp [hunk 2]
```diff
@@ -101,13 +125,22 @@ ddwaf_object* to_ddwaf_string(ddwaf_object *object, Napi::Value val, bool lim) {
...
-
+ddwaf_object* to_ddwaf_object(
+        ddwaf_object *object,
+        Napi::Env env,
+        Napi::Value val,
+        int depth,
+        bool lim
+) {
+  return to_ddwaf_object(object, env, val, depth, lim, false);
+}
```
Comment (simon-id) [inline src/convert.cpp:136]: can't we have a default value instead ? there: https://github.com/DataDog/dd-native-appsec-js/blob/main/src/convert.h#L19

File: src/convert.cpp [hunk 3]
```diff
@@ -101,13 +125,22 @@ ddwaf_object* to_ddwaf_string(ddwaf_object *object, Napi::Value val, bool lim) {
   }
   return ddwaf_object_stringl(object, str.c_str(), len);
 }
-
+ddwaf_object* to_ddwaf_object(
```
Comment (simon-id) [inline src/convert.cpp:128]:
```suggestion

ddwaf_object* to_ddwaf_object(
```

File: test/index.js [hunk 4]
```diff
@@ -842,11 +822,196 @@ describe('limit tests', () => {
...
+    const result1 = context1.run({
+      persistent: {
+        'server.request.body': body
+      }
+    }, TIMEOUT)
+    assert(!result1.status)
+
+    body.toJSON = function toJSON () {
+      assert(this === body)
+      return [{ a: '.htaccess' }]
```
Comment (simon-id) [inline test/index.js:876]:
```suggestion
      return ['.htaccess']
```

File: src/convert.cpp (hunk 2 above)
Comment (uurien) [reply src/convert.cpp:136]: ooh, I thought that it wasn't possible in c++

**Review by uurien — APPROVED** [review]

File: src/convert.cpp (hunk 2 above)
Comment (simon-id) [reply src/convert.cpp]: Aaah much cleaner like that perfect :+1:

**Review by simon-id — COMMENTED** [review]
LGTM

**Pull request merged**